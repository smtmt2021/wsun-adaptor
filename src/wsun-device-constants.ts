/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export const MAX_BUF_SIZE = 1024;
export const MAX_DGRAM_SIZE = 1024;

const isTesting = process.env.NODE_ENV === 'test';
export const COMMAND_TIMEOUT = (isTesting ? 1 : 3) * 1000; // msec
export const SEND_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const TERM_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const SCAN_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
export const CONNECTION_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
