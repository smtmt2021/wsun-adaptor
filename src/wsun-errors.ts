/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export class WsunError extends Error {
  constructor(message: string) {
    super('wsun: ' + message);
    this.name = new.target.name;
  }
}

export class WsunUdpTransmissionFailed extends WsunError {
  constructor() {
    super("wsun adaptor replied 'udp transmission failed'");
    this.name = new.target.name;
  }
}

export class WsunSendingInProgress extends WsunError {
  constructor() {
    super('wsun is sending in progress');
    this.name = new.target.name;
  }
}

export class WsunCommandError extends WsunError {
  private static reason2message(reason: number): string {
    switch (reason) {
      case 4:
        return "wsun adaptor replied 'unsupported command'";
      case 5:
        return "wsun adaptor replied 'incorrect command'";
      case 6:
        return "wsun adaptor replied 'incorrect range'";
      case 9:
        return "wsun adaptor replied 'uart error'";
      case 10:
        return "wsun adaptor replied 'failed to get result'";
      default:
        return "wsun adaptor replied 'unknown error'";
    }
  }

  constructor(public reason: number) {
    super(WsunCommandError.reason2message(reason));
    this.name = new.target.name;
  }
}

export class WsunTimeoutError extends WsunError {
  constructor(message: string) {
    super(message);
    this.name = new.target.name;
  }
}
