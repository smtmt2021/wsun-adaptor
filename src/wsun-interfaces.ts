/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { RemoteInfo } from 'dgram';

export type WsunEventUdp = 'rxudp';

export interface WsunCache {
  channel: number;
  panId: number;
  addr: string;
}

export interface WsunPanDesc extends WsunCache {
  page?: number;
  pairId?: string;
  LQI?: number;
}

export interface WsunId {
  id: string;
  password: string;
}

export interface WsunUDP {
  sender: string;
  dest: string;
  rport: number;
  lport: number;
  senderlla: string;
  secured: boolean;
  datalen: number;
  data: Buffer;
}

export interface WsunSocket {
  bind(port: number): Promise<void>;
  close(): Promise<void>;
  send(msg: Buffer | string, port: number, address: string): Promise<void>;
  on(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  once(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeAllListeners(event: 'message'): this;
}

export interface WsunConfig {
  device: string; // device name of serial port which is connected with wsun adaptor
}

export interface WsunAdaptorInternal {
  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Delete a used socket.
   *  This function is called automatically whenever a socket is closed.
   *  @param  socket  to be delete.
   */
  deleteSocket(socket: WsunSocket): void;

  /**
   *  Send a datagram message to the specified destination.
   *  This function is called via socket, do not call directly.
   *  @param  msg  to be sent.
   *  @param  port  is destination port.
   *  @param  address is destination address.
   */
  sendDgram(msg: Buffer, port: number, address: string): Promise<void>;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'rxudp'
   */
  on(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;

  /**
   * removes the specified listener from the event
   * @param en
   * @param listener
   */
  removeListener(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;
}
