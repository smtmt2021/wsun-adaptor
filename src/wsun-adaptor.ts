/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunId, WsunPanDesc, WsunSocket } from './';
import { WsunCache } from './wsun-interfaces';

export type WsunEventSession =
  | 'session-established'
  | 'session-ending'
  | 'session-ended'
  | 'session-error'
  | 'session-timeout'
  | 'session-108-start'
  | 'session-108-end';

export interface WsunAdaptor {
  /**
   *  Start activities with a wsun adaptor
   *  and establish a connection with PAA coordinator.
   *  @param  wsunId  ID & password
   *  @param  cache   cache info
   *  @returns  a descriptor of the connected PAA coordinator,
   *            undefined if no PAA coordinator was found.
   */
  open(wsunId: WsunId, cache?: WsunCache): Promise<WsunPanDesc | undefined>;

  /**
   *  Terminate a connection with a PAA coordinator
   */
  close(): Promise<void>;

  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Create a socket to communicate remote node.
   *  @return a socket.
   */
  createSocket(): WsunSocket;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'session-established' a PANA session is established
   * @event 'session-ending'      receive a PANA session end request
   * @event 'session-ended'       a PANA session is ended
   * @event 'session-error'       a PANA session is ended due to error
   * @event 'session-timeout'     a PANA session is ended due to timeout
   * @event 'session-108-start'   Restriction of ARIB 108 is started
   * @event 'session-108-end'     Restriction of ARIB 108 is ended
   */
  on(en: WsunEventSession, listener: () => void): void;
}
