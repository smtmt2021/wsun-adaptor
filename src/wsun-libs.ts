/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunCache } from './wsun-interfaces';

export const isParamOfPanDesc = /^ {2}[A-Za-z0-9 ]+:[A-Fa-f0-9]+$/;
export const isNumber = /^\d+$/;
export const isMac = /^[0-9A-Fa-f]{16}$/;
export const isIPv6 = /^[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+$/;

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function msec2duration(msec: number): number {
  msec = Math.max(msec, 20);
  const duration = Math.floor(Math.log2(msec / 10) - 1);
  return Math.min(duration, 14);
}

export function duration2msec(duration: number): number {
  return Math.pow(2, duration + 1) * 10;
}

export function isValidCache(cache: WsunCache): boolean {
  if (cache.channel < 33 || cache.channel > 60) {
    return false;
  }
  if (cache.panId < 0 || cache.panId > 0xffff) {
    return false;
  }
  if (!isMac.test(cache.addr)) {
    return false;
  }
  return true;
}
