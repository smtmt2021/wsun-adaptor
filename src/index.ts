/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunConfig } from './wsun-interfaces';
import { WsunAdaptor, WsunEventSession } from './wsun-adaptor';
import { WsunBp35a1 } from './wsun-bp35a1';
import DEBUG from 'debug';
const debug = DEBUG('wsun/adaptor');

export { MAX_DGRAM_SIZE } from './wsun-device-constants';
export { WsunAdaptor, WsunEventSession };
export { WsunId, WsunPanDesc, WsunSocket, WsunConfig, WsunCache } from './wsun-interfaces';
export {
  WsunError,
  WsunSendingInProgress,
  WsunUdpTransmissionFailed,
  WsunCommandError,
  WsunTimeoutError
} from './wsun-errors';

export type NameOfClass = 'BP35A1' | 'MOCK';
export interface MockConstructor {
  new (config: WsunConfig): WsunAdaptor;
}

let instance: WsunAdaptor;

/**
 *  Class factory to get a wsun adaptor
 *  @param  className is model name of adaptor you want
 *  @param  config
 *  @return an instance of adaptor which is specified by the model name
 */
export function WsunGetAdaptor(
  className: NameOfClass,
  config?: WsunConfig
): WsunAdaptor | undefined {
  debug('WsunGetAdaptor:', className, JSON.stringify(config));
  if (!instance && config) {
    switch (className) {
      case 'BP35A1':
        instance = new WsunBp35a1(config);
        break;
    }
  }
  return instance;
}
