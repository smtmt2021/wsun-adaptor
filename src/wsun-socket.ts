/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import promiseRetry from 'promise-retry';
import { EventEmitter } from 'events';
import { MAX_DGRAM_SIZE } from './wsun-device-constants';
import { WsunUDP, WsunSocket, WsunAdaptorInternal } from './wsun-interfaces';
import { isMac, isIPv6 } from './wsun-libs';
import { WsunError, WsunSendingInProgress } from './wsun-errors';
import DEBUG from 'debug';
const debug = DEBUG('wsun/socket');

function sleep(delay: number) {
  return new Promise<void>(resolve => setTimeout(() => resolve(), delay));
}

/**
 *  Datagram socket for communicating with wsun nodes,
 *  which provides some functionalities of Nodejs's UDP/datagram socket.
 */
export class WsunAdaptorSocket extends EventEmitter implements WsunSocket {
  private listeningPort?: number;
  private listener?: (udp: WsunUDP) => void;
  private busy = false;
  private closing = false;

  constructor(private adaptor: WsunAdaptorInternal) {
    super();
  }

  /**
   *  Listen for datagram messages on the specified port.
   *
   *  @param port - port number to be listened
   *  @event message - is fired when a new datagram is available on a socket.
   *    The event handler function is passed two arguments:
   *      `msg: Buffer` and `rinfo: RemoteInfo`.
   */
  bind(port: number): Promise<void> {
    debug('wsun-socket.bind', port);
    if (port < 1 || port > 0xffff) {
      return Promise.reject(new RangeError('The port muse be 1 <= and <= 0xffff'));
    }
    if (this.listeningPort) {
      return Promise.reject(new WsunError('Already listening the port'));
    }
    this.listeningPort = port;
    this.adaptor.on(
      'rxudp',
      (this.listener = (udp: WsunUDP) => {
        if (this.listeningPort === udp.rport) {
          this.emit('message', udp.data, {
            address: udp.sender,
            family: 'IPv6',
            port: udp.lport,
            size: udp.datalen
          });
        }
      })
    );
    return Promise.resolve();
  }

  /**
   * Close the underlying socket and stop listening for data on it.
   */
  async close(): Promise<void> {
    debug('wsun-socket.close');
    this.closing = true;
    if (this.busy) {
      console.info('wsun: closing socket is pending');
      await sleep(1 * 1000);
      return this.close();
    } else {
      if (this.listeningPort) {
        this.listeningPort = undefined;
      }
      if (this.listener) {
        this.adaptor.removeListener('rxudp', this.listener);
        this.listener = undefined;
      }
      this.adaptor.deleteSocket(this);
      this.closing = false;
      return Promise.resolve();
    }
  }

  /**
   *  Send a datagram message to the specified destination.
   *  @param  msg  to be sent.
   *  @param  port  is destination port.
   *  @param  address is destination address.
   */
  send(msg: Buffer, port: number, address: string): Promise<void> {
    debug('wsun-socket.send');
    if (port < 1 || port > 0xffff) {
      return Promise.reject(new RangeError('The port muse be 1 <= and <= 0xffff'));
    }
    if (!isMac.test(address) && !isIPv6.test(address)) {
      return Promise.reject(new TypeError(`${address} must be ipv6 or mac address`));
    }

    if (msg.length > MAX_DGRAM_SIZE) {
      return Promise.reject(new Error(`Size of dgram is too large, ${msg.length}`));
    }
    if (!this.adaptor.isOnline() || this.closing) {
      const err = new Error('Invalid state');
      console.error('wsadaptor: wsun-socket.send', err.stack);
      return Promise.reject(err);
    }

    this.busy = true;
    return promiseRetry(
      (retry, count) =>
        this.adaptor.sendDgram(msg, port, address).catch(err => {
          if (err instanceof WsunSendingInProgress) {
            console.info('wsadaptor: wssocket.send: retry', count);
            return retry(err);
          }
          throw err;
        }),
      {
        retries: 3
      }
    ).then(() => {
      this.busy = false;
    });
  }
}
