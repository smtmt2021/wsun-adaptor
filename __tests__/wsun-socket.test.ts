/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { EventEmitter } from 'events';
import * as WDC from '../src/wsun-device-constants';
import { WsunSocket, WsunEventUdp, WsunUDP } from '../src/wsun-interfaces';
import { WsunAdaptorSocket } from '../src/wsun-socket';
import { WsunError, WsunSendingInProgress } from '../src/wsun-errors';

interface WsunSocketSendMock {
  sendDgram: jest.Mock;
  isOnline: jest.Mock;
  deleteSocket: jest.Mock;
  on: jest.Mock;
  removeListener: jest.Mock;
}

function wsunSocketSendMock(): WsunSocketSendMock {
  return {
    sendDgram: jest.fn((msg: Buffer, port: number, address: string) => Promise.resolve()),
    isOnline: jest.fn(() => false),
    deleteSocket: jest.fn((socket: WsunSocket) => { }),
    on: jest.fn((en: WsunEventUdp, listener: (udp: WsunUDP) => void) => { }),
    removeListener: jest.fn((en: WsunEventUdp, listener: (udp: WsunUDP) => void) => { })
  };
}

let mock: WsunSocketSendMock;

describe('WsunSocket.send', () => {
  beforeEach(() => {
    mock = wsunSocketSendMock();
    mock.isOnline.mockReturnValue(true);
    mock.sendDgram.mockResolvedValue(undefined);
  });

  test('throw INVALID_RANGE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf = Buffer.from('Hello world!');
    const address = '1234567890123456';
    const port = 0;
    return expect(socket.send(buf, port, address)).rejects.toMatchSnapshot();
  });

  test('throw INVALID_RANGE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf = Buffer.from('Hello world!');
    const address = '1234567890123456';
    const port = 0x10000;
    return expect(socket.send(buf, port, address)).rejects.toMatchSnapshot();
  });

  test('throw INVALID_ADDRESS', () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf = Buffer.from('Hello world!');
    const port = 1;
    const address = '1234:2345:3456:4567:5678:6789:789A:89gh';
    return expect(socket.send(buf, port, address)).rejects.toMatchSnapshot();
  });

  test('throw TOO_LONG', () => {
    const socket = new WsunAdaptorSocket(mock);
    const port = 1;
    const address = '1234567890123456';
    const buf = Buffer.alloc(WDC.MAX_BUF_SIZE + 1);
    return expect(socket.send(buf, port, address)).rejects.toMatchSnapshot();
  });

  test('throw NOT_ONLINE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf = Buffer.from('Hello world!');
    const port = 1;
    const address = '1234567890123456';
    mock.isOnline.mockReturnValueOnce(false);
    return expect(socket.send(buf, port, address)).rejects.toMatchSnapshot();
  });

  test('throw NO_HANDLE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const port = 1;
    const address = '1234567890123456';
    const buf = Buffer.from('Hello world!');
    const error = new WsunError('wsun: no handle to send a dgram');
    mock.sendDgram.mockImplementationOnce(() => Promise.reject(error));
    return expect(socket.send(buf, port, address)).rejects.toEqual(error);
  });

  test('send successfully', () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf = Buffer.from('Hello world!');
    const port = 1;
    const address = '1234567890123456';
    return socket.send(buf, port, address).then(resp => {
      expect(resp).toEqual(undefined);
      expect(mock.sendDgram.mock.calls.length).toBe(1);
      expect(mock.sendDgram.mock.calls[0][0]).toEqual(buf);
      expect(mock.sendDgram.mock.calls[0][1]).toBe(port);
      expect(mock.sendDgram.mock.calls[0][2]).toBe(address);
    });
  });
});

describe('WsunSocket.bind', () => {
  beforeEach(() => {
    mock = wsunSocketSendMock();
    mock.isOnline.mockReturnValue(true);
    mock.sendDgram.mockResolvedValue(undefined);
  });

  test('throw INVALID_RANGE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const port = 0;
    return expect(socket.bind(port)).rejects.toMatchSnapshot();
  });

  test('throw INVALID_RANGE', () => {
    const socket = new WsunAdaptorSocket(mock);
    const port = 0x10000;
    return expect(socket.bind(port)).rejects.toMatchSnapshot();
  });

  test('bind successfully and bind again', async () => {
    const port = 1;
    const socket = new WsunAdaptorSocket(mock);
    await expect(socket.bind(port)).resolves.toBeUndefined();
    expect(mock.on.mock.calls.length).toBe(1);
    expect(mock.on.mock.calls[0].length).toBe(2);
    expect(mock.on.mock.calls[0][0]).toMatch('rxudp');

    await expect(socket.bind(port)).rejects.toMatchSnapshot();

    await expect(socket.close()).resolves.toBeUndefined();
    expect(mock.removeListener.mock.calls.length).toBe(1);
    expect(mock.removeListener.mock.calls[0].length).toBe(2);
    expect(mock.removeListener.mock.calls[0][0]).toMatch('rxudp');
    expect(typeof mock.removeListener.mock.calls[0][1] === 'function').toBeTruthy();
  });

  test('receive data', done => {
    const remotePort = 1;
    const localPort = 2;
    const socket = new WsunAdaptorSocket(mock);
    const emitter = new EventEmitter();
    const obuf = Buffer.from('Hello world!');
    const oudp = {
      sender: '1234:2345:3456:4567:5678:6789:789A:0001',
      dest: '1234:2345:3456:4567:5678:6789:789A:0002',
      rport: remotePort,
      lport: localPort,
      senderlla: '1122334455667788',
      secured: true,
      datalen: obuf.length,
      data: obuf
    };
    mock.on.mockImplementation((n: string, f: (rudp: WsunUDP) => void) => {
      emitter.on(n, f);
    });
    socket.once('message', (rudp, rinfo) => {
      expect(rudp).toEqual(obuf);
      expect(rinfo).toEqual({
        address: oudp.sender,
        family: 'IPv6',
        port: oudp.lport,
        size: oudp.datalen
      });
      done();
    });
    socket.bind(remotePort).then(() => emitter.emit('rxudp', oudp));
  });

  test('receive no data (port numbers are different)', done => {
    const remotePort = 1;
    const localPort = 2;
    const socket = new WsunAdaptorSocket(mock);
    const emitter = new EventEmitter();
    const obuf = Buffer.from('Hello world!');
    const oudp = {
      sender: '1234:2345:3456:4567:5678:6789:789A:0001',
      dest: '1234:2345:3456:4567:5678:6789:789A:0002',
      rport: remotePort,
      lport: localPort,
      senderlla: '1122334455667788',
      secured: true,
      datalen: obuf.length,
      data: obuf
    };
    mock.on.mockImplementation((n: string, f: (rudp: WsunUDP) => void) => {
      emitter.on(n, f);
    });
    const callback = jest.fn();
    setTimeout(() => {
      socket.removeAllListeners('message');
      expect(callback).not.toBeCalled();
      done();
    }, 1 * 1000);
    socket.once('message', callback);
    socket
      .bind(remotePort + 12345) // listening with different number
      .then(() => emitter.emit('rxudp', oudp));
  });
});

describe('WsunSocket.send 2 packets simultaneously', () => {
  beforeEach(() => {
    mock = wsunSocketSendMock();
    mock.isOnline.mockReturnValue(true);
    mock.sendDgram.mockResolvedValue(undefined);
    mock.sendDgram.mockImplementationOnce(
      () =>
        // for first call
        new Promise<void>(resolve => setTimeout(() => resolve(), 1 * 1000))
    );
    mock.sendDgram.mockImplementationOnce(() =>
      // for secound call
      Promise.reject(new WsunSendingInProgress())
    );
  });

  test('send 2 packets successfully', async () => {
    const socket = new WsunAdaptorSocket(mock);
    const buf1 = Buffer.from('Hello world!');
    const buf2 = Buffer.from('Hello another world!');
    const port = 1;
    const address = '1234567890123456';

    const promise = socket.send(buf1, port, address); // will be resolved after 1 sec.
    await expect(socket.send(buf2, port, address)).resolves.not.toThrow();
    await expect(promise).resolves.not.toThrow();

    expect(mock.sendDgram.mock.calls.length).toBe(3);
    expect(mock.sendDgram.mock.calls[0][0]).toEqual(buf1);
    expect(mock.sendDgram.mock.calls[0][1]).toBe(port);
    expect(mock.sendDgram.mock.calls[0][2]).toBe(address);

    expect(mock.sendDgram.mock.calls[1][0]).toEqual(buf2);
    expect(mock.sendDgram.mock.calls[1][1]).toBe(port);
    expect(mock.sendDgram.mock.calls[1][2]).toBe(address);

    expect(mock.sendDgram.mock.calls[2][0]).toEqual(buf2);
    expect(mock.sendDgram.mock.calls[2][1]).toBe(port);
    expect(mock.sendDgram.mock.calls[2][2]).toBe(address);
  });
});
