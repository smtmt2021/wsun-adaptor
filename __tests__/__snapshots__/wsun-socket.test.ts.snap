// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WsunSocket.bind bind successfully and bind again 1`] = `[WsunError: wsun: Already listening the port]`;

exports[`WsunSocket.bind throw INVALID_RANGE 1`] = `[RangeError: The port muse be 1 <= and <= 0xffff]`;

exports[`WsunSocket.bind throw INVALID_RANGE 2`] = `[RangeError: The port muse be 1 <= and <= 0xffff]`;

exports[`WsunSocket.send throw INVALID_ADDRESS 1`] = `[TypeError: 1234:2345:3456:4567:5678:6789:789A:89gh must be ipv6 or mac address]`;

exports[`WsunSocket.send throw INVALID_RANGE 1`] = `[RangeError: The port muse be 1 <= and <= 0xffff]`;

exports[`WsunSocket.send throw INVALID_RANGE 2`] = `[RangeError: The port muse be 1 <= and <= 0xffff]`;

exports[`WsunSocket.send throw NOT_ONLINE 1`] = `[Error: Invalid state]`;

exports[`WsunSocket.send throw TOO_LONG 1`] = `[Error: Size of dgram is too large, 1025]`;
